std::atomic_*() overloads for shared_ptr are deprecated in C++20

Just a heads-up that following functionality is deprecated in C++ 20:

```
D:\a\gatling\gatling\extern\openexr\src\lib\IlmThread\IlmThreadPool.cpp(119,52): warning C4996: 'std::atomic_load': warning STL4029: std::atomic_*() overloads for shared_ptr are deprecated in C++20. The shared_ptr specialization of std::atomic provides superior functionality. You can define _SILENCE_CXX20_OLD_SHARED_PTR_ATOMIC_SUPPORT_DEPRECATION_WARNING or _SILENCE_ALL_CXX20_DEPRECATION_WARNINGS to acknowledge that you have received this warning. [D:\a\gatling\gatling\BUILD\extern\openexr\src\lib\IlmThread\IlmThread.vcxproj]
D:\a\gatling\gatling\extern\openexr\src\lib\IlmThread\IlmThreadPool.cpp(123,33): warning C4996: 'std::atomic_exchange': warning STL4029: std::atomic_*() overloads for shared_ptr are deprecated in C++20. The shared_ptr specialization of std::atomic provides superior functionality. You can define _SILENCE_CXX20_OLD_SHARED_PTR_ATOMIC_SUPPORT_DEPRECATION_WARNING or _SILENCE_ALL_CXX20_DEPRECATION_WARNINGS to acknowledge that you have received this warning. [D:\a\gatling\gatling\BUILD\extern\openexr\src\lib\IlmThread\IlmThread.vcxproj]
  part.c
C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\VC\Tools\MSVC\14.29.30133\include\memory(3717,17): warning C4996: 'std::atomic_load_explicit': warning STL4029: std::atomic_*() overloads for shared_ptr are deprecated in C++20. The shared_ptr specialization of std::atomic provides superior functionality. You can define _SILENCE_CXX20_OLD_SHARED_PTR_ATOMIC_SUPPORT_DEPRECATION_WARNING or _SILENCE_ALL_CXX20_DEPRECATION_WARNINGS to acknowledge that you have received this warning. [D:\a\gatling\gatling\BUILD\extern\openexr\src\lib\IlmThread\IlmThread.vcxproj]
D:\a\gatling\gatling\extern\openexr\src\lib\IlmThread\IlmThreadPool.cpp(119): message : see reference to function template instantiation 'std::shared_ptr<IlmThread_3_2::ThreadPoolProvider> std::atomic_load<IlmThread_3_2::ThreadPoolProvider>(const std::shared_ptr<IlmThread_3_2::ThreadPoolProvider> *)' being compiled [D:\a\gatling\gatling\BUILD\extern\openexr\src\lib\IlmThread\IlmThread.vcxproj]
C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\VC\Tools\MSVC\14.29.30133\include\memory(3747,17): warning C4996: 'std::atomic_exchange_explicit': warning STL4029: std::atomic_*() overloads for shared_ptr are deprecated in C++20. The shared_ptr specialization of std::atomic provides superior functionality. You can define _SILENCE_CXX20_OLD_SHARED_PTR_ATOMIC_SUPPORT_DEPRECATION_WARNING or _SILENCE_ALL_CXX20_DEPRECATION_WARNINGS to acknowledge that you have received this warning. [D:\a\gatling\gatling\BUILD\extern\openexr\src\lib\IlmThread\IlmThread.vcxproj]
D:\a\gatling\gatling\extern\openexr\src\lib\IlmThread\IlmThreadPool.cpp(123): message : see reference to function template instantiation 'std::shared_ptr<IlmThread_3_2::ThreadPoolProvider> std::atomic_exchange<IlmThread_3_2::ThreadPoolProvider>(std::shared_ptr<IlmThread_3_2::ThreadPoolProvider> *,std::shared_ptr<IlmThread_3_2::ThreadPoolProvider>)' being compiled [D:\a\gatling\gatling\BUILD\extern\openexr\src\lib\IlmThread\IlmThread.vcxproj]
```

--- a/src/lib/IlmThread/IlmThreadPool.cpp
+++ b/src/lib/IlmThread/IlmThreadPool.cpp
@@ -120,18 +120,49 @@ struct ThreadPool::Data
     ~Data ();
     Data (const Data&)            = delete;
     Data& operator= (const Data&) = delete;
-    Data (Data&&)                 = default;
-    Data& operator= (Data&&)      = delete;
 
-    ProviderPtr getProvider () const { return std::atomic_load (&_provider); }
+#    if defined(__cpp_lib_atomic_shared_ptr) &&                                \
+        (__cpp_lib_atomic_shared_ptr >= 201711L)
+    Data (Data&& other) noexcept
+    {
+        ProviderPtr p = other._provider.load (std::memory_order_acquire);
+        _provider.store (p, std::memory_order_release);
+        other._provider.store (ProviderPtr{}, std::memory_order_release);
+    }
+#    else
+    Data (Data&&) = default;
+#    endif
+
+    Data& operator= (Data&&) = delete;
+
+    ProviderPtr getProvider () const
+    {
+#    if defined(__cpp_lib_atomic_shared_ptr) &&                                \
+        (__cpp_lib_atomic_shared_ptr >= 201711L)
+        return _provider.load (std::memory_order_acquire);
+#    else
+        return std::atomic_load (&_provider);
+#    endif
+    }
 
     void setProvider (ProviderPtr provider)
     {
+#    if defined(__cpp_lib_atomic_shared_ptr) &&                                \
+        (__cpp_lib_atomic_shared_ptr >= 201711L)
+        ProviderPtr curp =
+            _provider.exchange (provider, std::memory_order_acq_rel);
+#    else
         ProviderPtr curp = std::atomic_exchange (&_provider, provider);
+#    endif
         if (curp && curp != provider) curp->finish ();
     }
 
-    std::shared_ptr<ThreadPoolProvider> _provider;
+#    if defined(__cpp_lib_atomic_shared_ptr) &&                                \
+        (__cpp_lib_atomic_shared_ptr >= 201711L)
+    std::atomic<ProviderPtr> _provider;
+#    else
+    ProviderPtr _provider;
+#    endif
 };
 
 namespace
@@ -468,9 +499,13 @@ ThreadPool::Data::Data ()
 }
 
 ThreadPool::Data::Data (ThreadPoolProvider *p)
-    : _provider (p)
 {
-    // empty
+#    if defined(__cpp_lib_atomic_shared_ptr) &&                                \
+        (__cpp_lib_atomic_shared_ptr >= 201711L)
+    _provider.store (ProviderPtr (p), std::memory_order_release);
+#    else
+    _provider = ProviderPtr (p);
+#    endif
 }
 
 ThreadPool::Data::~Data ()

--- a/src/lib/IlmThread/IlmThreadPool.cpp
+++ b/src/lib/IlmThread/IlmThreadPool.cpp
@@ -35,6 +35,11 @@ using namespace oneapi;
 #    endif
 #endif
 
+#if defined(__cpp_lib_atomic_shared_ptr) &&                                \
+    (__cpp_lib_atomic_shared_ptr >= 201711L)
+#    define ILMTHREAD_HAVE_ATOMIC_SHARED_PTR
+#endif
+
 ILMTHREAD_INTERNAL_NAMESPACE_SOURCE_ENTER
 
 namespace
@@ -121,13 +126,12 @@ struct ThreadPool::Data
     Data (const Data&)            = delete;
     Data& operator= (const Data&) = delete;
 
-#    if defined(__cpp_lib_atomic_shared_ptr) &&                                \
-        (__cpp_lib_atomic_shared_ptr >= 201711L)
+#    ifdef ILMTHREAD_HAVE_ATOMIC_SHARED_PTR
     Data (Data&& other) noexcept
     {
-        ProviderPtr p = other._provider.load (std::memory_order_acquire);
+        ProviderPtr p =
+            other._provider.exchange (ProviderPtr{}, std::memory_order_acq_rel);
         _provider.store (p, std::memory_order_release);
-        other._provider.store (ProviderPtr{}, std::memory_order_release);
     }
 #    else
     Data (Data&&) = default;
@@ -137,8 +141,7 @@ struct ThreadPool::Data
 
     ProviderPtr getProvider () const
     {
-#    if defined(__cpp_lib_atomic_shared_ptr) &&                                \
-        (__cpp_lib_atomic_shared_ptr >= 201711L)
+#    ifdef ILMTHREAD_HAVE_ATOMIC_SHARED_PTR
         return _provider.load (std::memory_order_acquire);
 #    else
         return std::atomic_load (&_provider);
@@ -147,8 +150,7 @@ struct ThreadPool::Data
 
     void setProvider (ProviderPtr provider)
     {
-#    if defined(__cpp_lib_atomic_shared_ptr) &&                                \
-        (__cpp_lib_atomic_shared_ptr >= 201711L)
+#    ifdef ILMTHREAD_HAVE_ATOMIC_SHARED_PTR
         ProviderPtr curp =
             _provider.exchange (provider, std::memory_order_acq_rel);
 #    else
@@ -157,8 +159,7 @@ struct ThreadPool::Data
         if (curp && curp != provider) curp->finish ();
     }
 
-#    if defined(__cpp_lib_atomic_shared_ptr) &&                                \
-        (__cpp_lib_atomic_shared_ptr >= 201711L)
+#    ifdef ILMTHREAD_HAVE_ATOMIC_SHARED_PTR
     std::atomic<ProviderPtr> _provider;
 #    else
     ProviderPtr _provider;
@@ -500,8 +501,7 @@ ThreadPool::Data::Data ()
 
 ThreadPool::Data::Data (ThreadPoolProvider *p)
 {
-#    if defined(__cpp_lib_atomic_shared_ptr) &&                                \
-        (__cpp_lib_atomic_shared_ptr >= 201711L)
+#    ifdef ILMTHREAD_HAVE_ATOMIC_SHARED_PTR
     _provider.store (ProviderPtr (p), std::memory_order_release);
 #    else
     _provider = ProviderPtr (p);

